### PR TITLE
RTP uplifting frame reordering fixes

### PR DESCRIPTION
**Pull request by gpolitis**
I was analyzing a pcap file and I noticed re-order frames which the bridge mis-handles and causes it to break the RTP stream.

Say you send frames A B C and each is comprised of 3 RTP packets (3 is not important, it can be N). So the sender emits A1, A2, A3, B1, B2, B3, C1, C2, C3. If the bridge receives something like this A1, A2, A3, C1, B1, B2, B3, C2, C3 it will break the RTP stream because it will uplift the B frame to Timestamp_C+1.

tl;dr; the RTP uplifting logic can cope with some packet re-ordering, but it doesn’t expect whole frames to be re-ordered. I’ve confirmed that it can cause long freezes; I’m not sure about scrambling but I find it much less likely.

### CODE REVIEW

**Comment by ibauersachs** [comment]
I might have missed some discussion, but the change int -> long in the SSRCs changes a lot of public API. I understand the motivation (and hate once again Java for not having uints), but is this really necessary? At some point much care was taken to NOT use longs and things refactored from long to int...

**Comment by gpolitis** [comment]
@ibauersachs thanks for chiming in. I'm afraid of regressions so there's a good chance I'll take that back. I'm not aware of any decision to use ints instead of longs (fmj favors ints for sure) and it seems we're in favor of longs in LJ (at least in the past couple of years). I have reviewed a lot of PRs from @bgrozev and he's using them whenever possible. cc @lyubomir

**Comment by gpolitis** [comment]
jenkins it's ok to test

File: src/org/jitsi/impl/neomedia/transform/rewriting/SsrcRewriter.java [hunk 1]
```diff
@@ -106,6 +105,11 @@ protected boolean removeEldestEntry(Map.Entry eldest) {
         }
     };
 
+    /*
+     * 
+     */
+    private TimestampEntry maxSourceTsEntry;
```
Comment (damencho) [inline src/org/jitsi/impl/neomedia/transform/rewriting/SsrcRewriter.java]: Missing javadoc :)

File: src/org/jitsi/impl/neomedia/transform/rewriting/SsrcRewriter.java [hunk 2]
```diff
@@ -287,7 +364,8 @@ void rewriteTimestamp(RawPacket p)
                     + " to " + newValue);
             }
 
-            tsHistory.put(oldValue, new TimestampEntry(now, newValue));
+            tsHistory.put(oldValue, new TimestampEntry(now, oldValue, newValue));
+            maxSourceTsEntry = new TimestampEntry(now, oldValue, newValue);
         }
```
Comment (damencho) [inline src/org/jitsi/impl/neomedia/transform/rewriting/SsrcRewriter.java]: Duplicate object creation.

File: src/org/jitsi/impl/neomedia/transform/DiscardTransformEngine.java [hunk 3]
```diff
@@ -37,12 +39,33 @@
...
      * A map of source ssrc to {@link ResumableStreamRewriter}.
      */
     private final Map<Long, ResumableStreamRewriter> ssrcToRewriter
         = new HashMap<>();
 
     /**
+     *
+     */
+    private final MediaStream stream;
+
```
Comment (damencho) [inline src/org/jitsi/impl/neomedia/transform/DiscardTransformEngine.java]: Missing javadoc.

File: src/org/jitsi/impl/neomedia/transform/DiscardTransformEngine.java [hunk 4]
```diff
@@ -37,12 +39,33 @@
...
+    /**
+     *
+     * @param stream
+     */
+    public DiscardTransformEngine(MediaStream stream)
+    {
```
Comment (damencho) [inline src/org/jitsi/impl/neomedia/transform/DiscardTransformEngine.java]: Missing javadoc.

File: src/org/jitsi/impl/neomedia/rtcp/RemoteClockEstimator.java [hunk 5]
```diff
@@ -128,11 +140,15 @@ public void update(byte[] buf, int off, int len)
             }
         }
 
+
+        logger.debug("Updating the remote clock ssrc=" + ssrc
+            + ", systemTimeMs=" + systemTimeMs
+            + ", rtpTimestamp=" + rtptimestamp);
+
```
Comment (damencho) [inline src/org/jitsi/impl/neomedia/rtcp/RemoteClockEstimator.java]: Maybe add if (logger.isDebugEnabled()) to avoid string creation.

File: src/org/jitsi/impl/neomedia/rtp/ResumableStreamRewriter.java [hunk 6]
```diff
@@ -52,6 +52,8 @@
      */
     long highestTimestampSent = -1;
 
+    private int cycles = 0;
+
```
Comment (damencho) [inline src/org/jitsi/impl/neomedia/rtp/ResumableStreamRewriter.java]: javadoc missing
